// Copyright (c) 2018-2020 Thomas Minier
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "fr.univnantes.gdd.sage";
option java_outer_classname = "SageSPARQL";
option objc_class_prefix = "HLW";

package sage;

// The SaGe SPARQL query server
service SageSPARQL {
  // Execute a SPARQL query using the Web preemption model
  rpc Query (SageQuery) returns (SageResponse) {}
}

// The SPARQL query sent to the SaGe server
message SageQuery {
  string query = 1;
  string default_graph_uri = 2;
  string next_link = 3;
}

// A binding as per SPARQL specification
message Binding {
  string variable = 1;
  string value = 2;
}

message BindingSet {
  repeated Binding values = 1;
}

// The response to a SPARQL query
message SageResponse {
  repeated BindingSet bindings = 1;
  bool is_done = 2;
  string next_link = 3;
}